File: flutterapp/android/app/src/main/kotlin/com/example/flutterapp/GestureRecognizer.kt
package com.example.flutterapp

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.google.mediapipe.tasks.components.containers.NormalizedLandmark
import kotlin.math.sqrt

// 1. Data Model
data class GestureTemplate(
    val name: String,
    val fingerprint: List<Float>
)

class GestureRecognizer(private val context: Context) {

    private val templates = mutableListOf<GestureTemplate>()
    private val gson = Gson()
    private val PREFS_NAME = "GesturePrefs"
    private val KEY_GESTURES = "SavedGestures"

    // LOWER = Stricter match. 
    // 15.0 is a good starting point for normalized coordinates.
    private val MATCH_THRESHOLD = 15.0f 

    init {
        loadFromStorage()
    }

    // --- MAIN API ---

    fun saveTemplate(name: String, landmarks: List<NormalizedLandmark>) {
        val fingerprint = createFingerprint(landmarks)
        
        // Overwrite if name exists
        templates.removeAll { it.name == name }
        templates.add(GestureTemplate(name, fingerprint))
        
        Log.d("GestureRecognizer", "Saved '$name'. Total templates: ${templates.size}")
        saveToStorage()
    }

    fun recognize(landmarks: List<NormalizedLandmark>): String {
        if (templates.isEmpty()) return "UNKNOWN"

        val currentFingerprint = createFingerprint(landmarks)
        
        var bestMatchName = "UNKNOWN"
        var bestScore = Float.MAX_VALUE

        for (template in templates) {
            val score = calculateDistance(currentFingerprint, template.fingerprint)
            
            if (score < bestScore) {
                bestScore = score
                bestMatchName = template.name
            }
        }

        return if (bestScore < MATCH_THRESHOLD) bestMatchName else "UNKNOWN"
    }

    // --- STORAGE LOGIC ---

    private fun saveToStorage() {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val json = gson.toJson(templates)
        prefs.edit().putString(KEY_GESTURES, json).apply()
    }

    private fun loadFromStorage() {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val json = prefs.getString(KEY_GESTURES, null)
        if (json != null) {
            val type = object : TypeToken<MutableList<GestureTemplate>>() {}.type
            val savedList: MutableList<GestureTemplate> = gson.fromJson(json, type)
            templates.clear()
            templates.addAll(savedList)
        }
    }

    // --- MATH & NORMALIZATION ---

    private fun createFingerprint(landmarks: List<NormalizedLandmark>): List<Float> {
        val fingerprint = mutableListOf<Float>()
        
        // 1. Use Wrist (Index 0) as Origin
        val wrist = landmarks[0]

        // 2. Scale by hand size (Wrist to Middle Finger Tip)
        val middleTip = landmarks[12]
        val scale = getDistance(wrist, middleTip)

        // 3. Normalize all points
        for (landmark in landmarks) {
            fingerprint.add((landmark.x() - wrist.x()) / scale)
            fingerprint.add((landmark.y() - wrist.y()) / scale)
            fingerprint.add((landmark.z() - wrist.z()) / scale)
        }
        return fingerprint
    }

    private fun calculateDistance(listA: List<Float>, listB: List<Float>): Float {
        var sum = 0.0f
        for (i in listA.indices) {
            val diff = listA[i] - listB[i]
            sum += diff * diff
        }
        return sqrt(sum)
    }

    private fun getDistance(p1: NormalizedLandmark, p2: NormalizedLandmark): Float {
        val dx = p1.x() - p2.x()
        val dy = p1.y() - p2.y()
        val dz = p1.z() - p2.z()
        return sqrt(dx*dx + dy*dy + dz*dz)
    }

    fun getSavedGestureNames(): List<String> {
        return templates.map { it.name }
    }
}

File: flutterapp/ios/GestureRecognizer.swift
//
//  GestureRecognizer.swift
//  
//
//  Created by Okan Demirbilek on 14.12.25.
//

import Foundation
import MediaPipeTasksVision

struct GestureTemplate: Codable {
    let name: String
    let fingerprint: [Float]
}

class GestureRecognizer {

    private var templates: [GestureTemplate] = []
    private let storageURL: URL

    private let MATCH_THRESHOLD: Float = 15.0

    init() {
        // Storagepath
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        self.storageURL = dir.appendingPathComponent("gestures.json")
        loadFromStorage()
        if templates.isEmpty {
            print("iOS: No saved gestures yet → returning empty list.")
        }

    }

    // MAIN API

    func saveTemplate(name: String, landmarks: [NormalizedLandmark]) {
        let fp = createFingerprint(from: landmarks)

        templates.removeAll { $0.name == name }
        templates.append(GestureTemplate(name: name, fingerprint: fp))

        saveToStorage()
        print("GestureRecognizer.swift: Saved template '\(name)'")
    }

    func recognize(landmarks: [NormalizedLandmark]) -> String {
        guard !templates.isEmpty else { return "UNKNOWN" }

        let current = createFingerprint(from: landmarks)

        var bestName = "UNKNOWN"
        var bestScore = Float.greatestFiniteMagnitude

        for template in templates {
            let score = calculateDistance(a: current, b: template.fingerprint)
            if score < bestScore {
                bestScore = score
                bestName = template.name
            }
        }

        return bestScore < MATCH_THRESHOLD ? bestName : "UNKNOWN"
    }

    func getSavedNames() -> [String] {
        return templates.map { $0.name }
    }

    // STORAGE

    private func saveToStorage() {
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(templates)
            try FileManager.default.createDirectory(
                at: storageURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            try data.write(to: storageURL)
        } catch {
            print("GestureRecognizer.swift ERROR saving: \(error)")
        }
    }

    private func loadFromStorage() {
        do {
            let data = try Data(contentsOf: storageURL)
            let loaded = try JSONDecoder().decode([GestureTemplate].self, from: data)
            templates = loaded
            print("Loaded \(templates.count) gesture templates")
        } catch {
            print("GestureRecognizer.swift: No saved gestures yet")
        }
    }

    // FINGERPRINTING

    private func createFingerprint(from landmarks: [NormalizedLandmark]) -> [Float] {
        guard landmarks.count >= 21 else { return [] }

        var result: [Float] = []

        let wrist = landmarks[0]
        let middleTip = landmarks[12]

        let scale = distance(wrist, middleTip)

        for lm in landmarks {
            result.append((lm.x - wrist.x) / scale)
            result.append((lm.y - wrist.y) / scale)
            result.append((lm.z - wrist.z) / scale)
        }
        return result
    }

    private func distance(_ a: NormalizedLandmark, _ b: NormalizedLandmark) -> Float {
        let dx = a.x - b.x
        let dy = a.y - b.y
        let dz = a.z - b.z
        return sqrt(dx*dx + dy*dy + dz*dz)
    }

    private func calculateDistance(a: [Float], b: [Float]) -> Float {
        var sum: Float = 0
        for i in 0 ..< a.count {
            let diff = a[i] - b[i]
            sum += diff * diff
        }
        return sqrt(sum)
    }
}
